ext/thymeleaf/src/main/java/org/mvcspec/ozark/ext/thymeleaf/CDIWebContext.java

```diff
@@ -0,0 +1,145 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.mvcspec.ozark.ext.thymeleaf;
+
+import org.thymeleaf.context.IWebContext;
+import org.thymeleaf.context.LazyContextVariable;
+
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.util.AnnotationLiteral;
+import javax.servlet.ServletContext;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Thymeleaf context for resolving variables from the model or cdi context. In addition to the functions provided by
+ * {@link org.thymeleaf.context.WebContext} this allows to resolve beans annotated with {@linkplain javax.inject.Named}
+ * from the cdi context.
+ * <p>
+ * Beans will be reolved in this order:
+ * <ol>
+ *     <li>special variables (like "request")</li>
+ *     <li>variables from the model</li>
+ *     <li>named beans from the cdi context</li>
+ * </ol>
+ * </p>
+ *
+ * @author Gregor Tudan
+ */
+class CDIWebContext implements IWebContext {
+
+    private final HttpServletRequest request;
+    private final HttpServletResponse response;
+    private final ServletContext context;
+    private final Locale locale;
+
+    private final BeanManager beanManager;
+    private final Set<String> cdiNamedBeans;
+
+    private final Map<String, Object> variables = new HashMap<>();
+
+    CDIWebContext(BeanManager beanManager, HttpServletRequest request, HttpServletResponse response, ServletContext servletContext, Locale locale) {
+        this.beanManager = beanManager;
+        this.request = request;
+        this.response = response;
+        this.context = servletContext;
+        this.locale = locale;
+        this.cdiNamedBeans = enumerateNamedBeans();
+    }
+
+    @SuppressWarnings("serial")
+    private Set<String> enumerateNamedBeans() {
+        Set<Bean<?>> beans = beanManager.getBeans(Object.class, new AnnotationLiteral<Any>() { });
+        return beans.stream().map(Bean::getName).filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    @Override
+    public HttpServletRequest getRequest() {
+        return request;
+    }
+
+    @Override
+    public HttpServletResponse getResponse() {
+        return response;
+    }
+
+    @Override
+    public HttpSession getSession() {
+        return request.getSession();
```

Comment:
I think we can use `request.getSession(false)` here... It looks like `null` is allowed in this case... 

See the [API docs of IWebContext](https://www.thymeleaf.org/apidocs/thymeleaf/3.0.0.RELEASE/org/thymeleaf/context/IWebContext.html#getSession--).

Reply:
Right, done